yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnTimelineUtils.scala

```diff
@@ -0,0 +1,720 @@
+    }
+  }
+
+  /**
+   * Convert a spark event to a timeline event
+   * @param event handled spark event
+   * @return a timeline event
+   */
+  def toTimelineEvent(event: HandleSparkEvent): TimelineEvent = {
+    val tlEvent = new TimelineEvent()
+    tlEvent.setEventType(Utils.getFormattedClassName(event.sparkEvent)
+        + "-" + YarnTimelineUtils.uid.incrementAndGet.toString)
+    tlEvent.setTimestamp(event.time)
+    val kvMap = new JHashMap[String, Object]()
+    val json = JsonProtocol.sparkEventToJson(event.sparkEvent)
+    val jObject = json.asInstanceOf[JObject]
+    // the timeline event wants a map of java objects for Jackson to serialize
+    val hashMap = toJavaMap(jObject)
+    tlEvent.setEventInfo(hashMap)
+    tlEvent
+  }
+
+  /**
+   * Describe the event for logging
+   * @param event timeline event
+   * @return a description
+   */
+  def describeEvent(event: TimelineEvent): String = {
+    var sparkEventDetails = ""
+    try {
+      sparkEventDetails = toSparkEvent(event).toString
+    } catch {
+      case _: Exception =>
+        sparkEventDetails = "(cannot convert event details to spark exception)"
+    }
+    s"${event.getEventType()} @ ${new Date(event.getTimestamp())}" +
+        s"\n    ${sparkEventDetails}"
+  }
+
+  /**
+   * Create details of a timeline entity, by describing every event inside it
+   * @param entity entity containing a possibly empty or null list of events
+   * @return a list of event details, with a newline between each one
+   */
+  def eventDetails(entity: TimelineEntity): String = {
+    val events: util.List[TimelineEvent] = entity.getEvents
+    if (events != null) {
+      events.foldLeft("") {
+        (s, evt) => (s + "\n" + describeEvent(evt))
+      }
+    } else {
+      ""
+    }
+  }
+
+  /**
+   * Describe a timeline entity
+   * @param entity entity
+   * @return a string description.
+   */
+  def describeEntity(entity: TimelineEntity): String = {
+    val events: util.List[TimelineEvent] = entity.getEvents
+    val eventSummary = if (events != null) {
+      s"contains ${events.size()} event(s)"
+    } else {
+      "contains no events"
+    }
+
+    val header = s"${entity.getEntityType }/${entity.getEntityId } @${entity.getDomainId }}"
+    val otherInfo = entity.getOtherInfo().foldLeft("\n") (
+      (acc, kv) => acc + s" ${kv._1} = ${kv._2}; "
+    )
+    s"Timeline Entity " + header +
+        " " + otherInfo +
+        " " + timeFieldToString(entity.getStartTime(), "start") +
+        " " + eventSummary
+  }
+
+  /**
+   * Convert a `java.lang.Long` reference to a string value, or, if the reference is null,
+   * to text declaring that the named field is empty
+   * @param time time reference
+   * @param field field name for error message
+   * @return a string to describe the field
+   */
+  def timeFieldToString(time: lang.Long, field: String): String = {
+    if (time != null) {
+      new Date(time).toString
+    } else {
+       ("no " + field + " time")
```

Comment:
- indentation is off
- `s"no $field time"`

The final string reads a little weirdly, too.